InnerEye/ML/run_ml.py

```diff
@@ -356,30 +372,36 @@ def run(self) -> None:
-                should_wait_for_other_child_runs = (not self.is_offline_run) and \
-                                                   self.container.cross_validation_split_index == 0
-                if should_wait_for_other_child_runs:
-                    self.wait_for_runs_to_finish()
-                    self.create_ensemble_model_and_run_inference()
-        else:
-            # Inference for all models that are specified via LightningContainers.
-            with logging_section("Model inference"):
-                self.run_inference_for_lightning_models(self.checkpoint_handler.get_checkpoints_to_test())
-            # We can't enforce that files are written to the output folder, hence change the working directory manually
-            with change_working_directory(self.container.outputs_folder):
-                self.container.create_report()
+        # Register the model, and then run inference as required. No models should be registered when running outside AzureML.
+        if not self.is_offline_run:
+            if self.should_register_model():
+                self.register_model(self.checkpoint_handler, ModelProcessing.DEFAULT)
+
+        if not self.azure_config.only_register_model:
+            if isinstance(self.container, InnerEyeContainer):
+                # Inference for the InnerEye built-in models
+                # We specify the ModelProcessing as DEFAULT here even if the run_recovery points to an ensemble run, because
+                # the current run is a single one. See the documentation of ModelProcessing for more details.
+                self.run_inference(self.checkpoint_handler, ModelProcessing.DEFAULT)
+
+                if self.container.generate_report:
+                    self.generate_report(ModelProcessing.DEFAULT)
+
+                # If this is an cross validation run, and the present run is child run 0, then wait for the sibling runs,
+                # build the ensemble model, and write a report for that.
+                if self.container.number_of_cross_validation_splits > 0:
+                    should_wait_for_other_child_runs = (not self.is_offline_run) and \
+                                                       self.container.cross_validation_split_index == 0
+                    if should_wait_for_other_child_runs:
+                        self.wait_for_runs_to_finish()
+                        self.create_ensemble_model_and_run_inference()
+            else:
+                # Inference for all models that are specified via LightningContainers.
+                with logging_section("Model inference"):
+                    self.run_inference_for_lightning_models(self.checkpoint_handler.get_checkpoints_to_test())
+                # We can't enforce that files are written to the output folder, hence change the working directory manually
+                with change_working_directory(self.container.outputs_folder):
+                    self.container.create_report()
```

Comment:
we have an inconsistency here, for legacy models we have the "generate_report" flag, but we are not respecting that here. And why do we have the generate_report flag actually?

Reply:
For LightningContainers, if the user doesn't want to write out a report, all the user needs to do is leave the create_report function empty. That is not an option for legacy models, we need the generate_report flag to disable writing out a report when it is not needed, since generate_report is defined in MLRunner. Keeping the flag is a good idea - for some models, it may take quite a while to write the report out, and it may not make too much sense to write out the report.

Reply:
Opened an issue for this: #457.